ctru-rs/src/services/ir_user.rs

```diff
@@ -0,0 +1,504 @@
+                    bit_rate: IR_BITRATE,
+                    shared_memory_handle,
+                })?;
+
+                // Set up our service state
+                let user_state = IrUserState {
+                    service_handle,
+                    shared_memory_handle,
+                    shared_memory,
+                    shared_memory_layout,
+                    recv_buffer_size,
+                    recv_packet_count,
+                };
+                *IR_USER_STATE.lock().unwrap() = Some(user_state);
+
+                Ok(())
+            },
+            || {
+                // Remove our service state from the global location
+                let mut shared_mem_guard = IR_USER_STATE.lock().unwrap();
+                let shared_mem = shared_mem_guard.take().unwrap();
+
+                (move || unsafe {
+                    // Close service and memory handles
+                    ResultCode(ctru_sys::svcCloseHandle(shared_mem.service_handle))?;
+                    ResultCode(ctru_sys::svcCloseHandle(shared_mem.shared_memory_handle))?;
+
+                    // Free shared memory
+                    std::alloc::dealloc(
+                        shared_mem.shared_memory.as_ptr() as *mut u8,
+                        shared_mem.shared_memory_layout,
+                    );
+
+                    Ok(())
+                })()
+                .unwrap();
+            },
+        )?;
+
+        Ok(IrUser {
+            _service_reference: service_reference,
+        })
+    }
+
+    /// Try to connect to the device with the provided ID.
+    pub fn require_connection(&self, device_id: IrDeviceId) -> crate::Result<()> {
+        self.send_service_request(
+            vec![REQUIRE_CONNECTION_COMMAND_HEADER, device_id.get_id()],
+            2,
+        )?;
+        Ok(())
+    }
+
+    /// Close the current IR connection.
+    pub fn disconnect(&self) -> crate::Result<()> {
+        self.send_service_request(vec![DISCONNECT_COMMAND_HEADER], 2)?;
+        Ok(())
+    }
+
+    /// Get an event handle that activates on connection status changes.
+    pub fn get_connection_status_event(&self) -> crate::Result<Handle> {
+        let response =
+            self.send_service_request(vec![GET_CONNECTION_STATUS_EVENT_COMMAND_HEADER], 4)?;
+        let status_event = response[3] as Handle;
+
+        Ok(status_event)
+    }
+
+    /// Get an event handle that activates when a packet is received.
+    pub fn get_recv_event(&self) -> crate::Result<Handle> {
+        let response = self.send_service_request(vec![GET_RECEIVE_EVENT_COMMAND_HEADER], 4)?;
+        let recv_event = response[3] as Handle;
+
+        Ok(recv_event)
+    }
+
+    /// Wait for an event to fire. If the timeout is reached, an error is returned. You can use
+    /// [`Error::is_timeout`] to check if the error is due to a timeout.
+    pub fn wait_for_event(event: Handle, timeout: Duration) -> crate::Result<()> {
```

Comment:
This should probably be moved into a new `svc` module

Reply:
During the `cam` module addition, we deemed `svc` not well suited for user management yet. If you can find a way to make it completely safe, then we can work to re-write it.

Reply:
We can have a limited API surface, like just supporting `svcWaitSynchronization`

Reply:
That’s a good idea for now 👌

Reply:
Updated